File: python/equity/01-options-pricing/src/eq_options/greeks.py
"""Analytic Black-Scholes Greeks and finite-difference Greeks for any pricer.

Units and conventions
---------------------
* delta : dV/dS, dimensionless (per unit of spot).
* gamma : d2V/dS2, per currency unit.
* vega  : dV/dsigma, currency units per unit of annualised vol
  (divide by 100 for the market's 'per vol point').
* theta : dV/dt, currency units per *year* (divide by 365 for per-day).
* rho   : dV/dr, currency units per unit of rate (divide by 100 for per bp*100).
* vanna : d2V/(dS dsigma).
* volga : d2V/dsigma2 (a.k.a. vomma).

All rates continuously compounded, annualised; ``T`` in years (ACT/365F).
"""

from __future__ import annotations

import math
from dataclasses import dataclass, fields
from typing import Callable

from scipy.stats import norm

from .black_scholes import OptionType, bs_price, d1_d2, validate_inputs

__all__ = [
    "BSGreeks",
    "bs_greeks",
    "fd_greeks",
    "compare_greeks",
    "delta",
    "gamma",
    "vega",
    "theta",
    "rho",
    "vanna",
    "volga",
]

_SQRT_2PI = math.sqrt(2.0 * math.pi)

Pricer = Callable[..., float]


@dataclass(frozen=True)
class BSGreeks:
    """Container for the full Greek set of a European option.

    Attributes use the units documented in the module docstring.
    """

    price: float
    delta: float
    gamma: float
    vega: float
    theta: float
    rho: float
    vanna: float
    volga: float

    def as_dict(self) -> dict[str, float]:
        """Return the Greeks as a plain ``dict`` (field name -> value)."""
        return {f.name: getattr(self, f.name) for f in fields(self)}


def bs_greeks(
    S: float,
    K: float,
    T: float,
    r: float,
    sigma: float,
    q: float = 0.0,
    option_type: OptionType = "call",
) -> BSGreeks:
    """Analytic Black-Scholes-Merton Greeks (with continuous dividend yield).

    Parameters
    ----------
    S, K : float
        Spot and strike (currency units), strictly positive.
    T : float
        Time to expiry in years (ACT/365F), strictly positive.
    r : float
        Continuously compounded annualised risk-free rate.
    sigma : float
        Annualised volatility, strictly positive.
    q : float
        Continuously compounded annualised dividend yield.
    option_type : {"call", "put"}
        Option payoff direction.

    Returns
    -------
    BSGreeks
        Price plus delta, gamma, vega, theta, rho, vanna, volga.

    Raises
    ------
    ValueError
        If inputs are invalid or ``S``, ``K``, ``T``, ``sigma`` are not
        strictly positive (Greeks are singular at the boundary).
    """
    d1, d2 = d1_d2(S, K, T, r, sigma, q)
    sqrt_t = math.sqrt(T)
    df_q = math.exp(-q * T)
    df_r = math.exp(-r * T)
    pdf_d1 = math.exp(-0.5 * d1 * d1) / _SQRT_2PI

    gamma_ = df_q * pdf_d1 / (S * sigma * sqrt_t)
    vega_ = S * df_q * pdf_d1 * sqrt_t
    vanna_ = -df_q * pdf_d1 * d2 / sigma
    volga_ = vega_ * d1 * d2 / sigma
    common_theta = -S * df_q * pdf_d1 * sigma / (2.0 * sqrt_t)

    if option_type == "call":
        price = S * df_q * norm.cdf(d1) - K * df_r * norm.cdf(d2)
        delta_ = df_q * norm.cdf(d1)
        theta_ = common_theta + q * S * df_q * norm.cdf(d1) - r * K * df_r * norm.cdf(d2)
        rho_ = K * T * df_r * norm.cdf(d2)
    elif option_type == "put":
        price = K * df_r * norm.cdf(-d2) - S * df_q * norm.cdf(-d1)
        delta_ = -df_q * norm.cdf(-d1)
        theta_ = common_theta - q * S * df_q * norm.cdf(-d1) + r * K * df_r * norm.cdf(-d2)
        rho_ = -K * T * df_r * norm.cdf(-d2)
    else:
        raise ValueError(f"option_type must be 'call' or 'put', got {option_type!r}")

    return BSGreeks(
        price=price,
        delta=delta_,
        gamma=gamma_,
        vega=vega_,
        theta=theta_,
        rho=rho_,
        vanna=vanna_,
        volga=volga_,
    )


def delta(S: float, K: float, T: float, r: float, sigma: float, q: float = 0.0,
          option_type: OptionType = "call") -> float:
    """Analytic delta dV/dS (dimensionless). See :func:`bs_greeks`."""
    return bs_greeks(S, K, T, r, sigma, q, option_type).delta


def gamma(S: float, K: float, T: float, r: float, sigma: float, q: float = 0.0,
          option_type: OptionType = "call") -> float:
    """Analytic gamma d2V/dS2 (per currency unit). See :func:`bs_greeks`."""
    return bs_greeks(S, K, T, r, sigma, q, option_type).gamma


def vega(S: float, K: float, T: float, r: float, sigma: float, q: float = 0.0,
         option_type: OptionType = "call") -> float:
    """Analytic vega dV/dsigma (per unit of vol). See :func:`bs_greeks`."""
    return bs_greeks(S, K, T, r, sigma, q, option_type).vega


def theta(S: float, K: float, T: float, r: float, sigma: float, q: float = 0.0,
          option_type: OptionType = "call") -> float:
    """Analytic theta dV/dt (per year). See :func:`bs_greeks`."""
    return bs_greeks(S, K, T, r, sigma, q, option_type).theta


def rho(S: float, K: float, T: float, r: float, sigma: float, q: float = 0.0,
        option_type: OptionType = "call") -> float:
    """Analytic rho dV/dr (per unit of rate). See :func:`bs_greeks`."""
    return bs_greeks(S, K, T, r, sigma, q, option_type).rho


def vanna(S: float, K: float, T: float, r: float, sigma: float, q: float = 0.0,
          option_type: OptionType = "call") -> float:
    """Analytic vanna d2V/(dS dsigma). See :func:`bs_greeks`."""
    return bs_greeks(S, K, T, r, sigma, q, option_type).vanna


def volga(S: float, K: float, T: float, r: float, sigma: float, q: float = 0.0,
          option_type: OptionType = "call") -> float:
    """Analytic volga d2V/dsigma2. See :func:`bs_greeks`."""
    return bs_greeks(S, K, T, r, sigma, q, option_type).volga


def fd_greeks(
    pricer: Pricer,
    S: float,
    K: float,
    T: float,
    r: float,
    sigma: float,
    q: float = 0.0,
    option_type: OptionType = "call",
    rel_bump: float = 1e-5,
    rel_bump2: float = 2e-4,
    **pricer_kwargs: object,
) -> BSGreeks:
    """Central finite-difference Greeks for *any* pricer with the BS signature.

    The pricer must accept ``(S, K, T, r, sigma, q, option_type, **kwargs)``
    positionally/by keyword and return a float price. Central differences
    are used everywhere; second derivatives use the standard three-point
    stencil; vanna uses the four-point cross stencil. Theta is reported as
    ``dV/dt = -dV/dT`` (per year).

    Parameters
    ----------
    pricer : callable
        Pricing function, e.g. :func:`eq_options.black_scholes.bs_price`
        or a lambda around :func:`eq_options.binomial.crr_price`.
    S, K, T, r, sigma, q, option_type
        Contract and market inputs, as in :func:`bs_greeks`.
    rel_bump : float
        Relative bump size for first derivatives; absolute bumps are
        ``rel_bump * max(|x|, 1)``. The default ``1e-5`` balances
        truncation vs round-off error for analytic pricers; noisy pricers
        (MC) need larger bumps.
    rel_bump2 : float
        Relative bump for second derivatives (gamma, vanna, volga), where
        round-off scales like ``eps / h^2`` and needs a larger ``h``
        (optimal ``h ~ eps^0.25``).
    **pricer_kwargs
        Extra keyword arguments forwarded to ``pricer`` (e.g. ``n_steps``).

    Returns
    -------
    BSGreeks
        Finite-difference price and Greeks (vanna/volga included).

    Raises
    ------
    ValueError
        If inputs are invalid, or ``T`` is too small to bump centrally.
    """
    validate_inputs(S, K, T, sigma, option_type)

    def f(s: float = S, sig: float = sigma, t: float = T, rr: float = r) -> float:
        return float(pricer(s, K, t, rr, sig, q, option_type, **pricer_kwargs))

    h_s = rel_bump * max(abs(S), 1.0)
    h_v = rel_bump * max(abs(sigma), 1.0)
    h_t = rel_bump * max(abs(T), 1.0)
    h_r = rel_bump * max(abs(r), 1.0)
    h_s2 = rel_bump2 * max(abs(S), 1.0)
    h_v2 = rel_bump2 * max(abs(sigma), 1.0)
    if T - h_t <= 0.0:
        raise ValueError(f"T={T} too small for a central theta bump of {h_t}")

    price = f()
    delta_ = (f(s=S + h_s) - f(s=S - h_s)) / (2.0 * h_s)
    gamma_ = (f(s=S + h_s2) - 2.0 * price + f(s=S - h_s2)) / (h_s2 * h_s2)
    vega_ = (f(sig=sigma + h_v) - f(sig=sigma - h_v)) / (2.0 * h_v)
    theta_ = -(f(t=T + h_t) - f(t=T - h_t)) / (2.0 * h_t)
    rho_ = (f(rr=r + h_r) - f(rr=r - h_r)) / (2.0 * h_r)
    vanna_ = (
        f(s=S + h_s2, sig=sigma + h_v2)
        - f(s=S + h_s2, sig=sigma - h_v2)
        - f(s=S - h_s2, sig=sigma + h_v2)
        + f(s=S - h_s2, sig=sigma - h_v2)
    ) / (4.0 * h_s2 * h_v2)
    volga_ = (f(sig=sigma + h_v2) - 2.0 * price + f(sig=sigma - h_v2)) / (h_v2 * h_v2)

    return BSGreeks(
        price=price,
        delta=delta_,
        gamma=gamma_,
        vega=vega_,
        theta=theta_,
        rho=rho_,
        vanna=vanna_,
        volga=volga_,
    )


def compare_greeks(
    S: float,
    K: float,
    T: float,
    r: float,
    sigma: float,
    q: float = 0.0,
    option_type: OptionType = "call",
    rel_bump: float = 1e-5,
) -> dict[str, dict[str, float]]:
    """Tabulate analytic vs central finite-difference Black-Scholes Greeks.

    Parameters
    ----------
    S, K, T, r, sigma, q, option_type
        As in :func:`bs_greeks`.
    rel_bump : float
        Relative bump used for the finite-difference leg.

    Returns
    -------
    dict
        ``{greek: {"analytic": x, "finite_diff": y, "abs_err": |x-y|}}``
        for each Greek including price.
    """
    ana = bs_greeks(S, K, T, r, sigma, q, option_type).as_dict()
    num = fd_greeks(bs_price, S, K, T, r, sigma, q, option_type, rel_bump).as_dict()
    return {
        name: {
            "analytic": ana[name],
            "finite_diff": num[name],
            "abs_err": abs(ana[name] - num[name]),
        }
        for name in ana
    }

File: rust/equity-options-engine/src/greeks.rs
//! Analytic Black-Scholes Greeks and finite-difference Greeks for any pricer.
//!
//! # Units and conventions (identical to the Python reference)
//!
//! * `delta` : dV/dS, dimensionless (per unit of spot).
//! * `gamma` : d2V/dS2, per currency unit.
//! * `vega`  : dV/dsigma, currency units per **unit** of annualised vol
//!   (divide by 100 for the market's "per vol point").
//! * `theta` : dV/dt, currency units per **year** (divide by 365 for per-day).
//! * `rho`   : dV/dr, currency units per **unit** of rate.
//! * `vanna` : d2V/(dS dsigma).
//! * `volga` : d2V/dsigma2 (a.k.a. vomma).
//!
//! All rates continuously compounded, annualised; `t` in years (ACT/365F).

use crate::black_scholes::{
    d1_d2, norm_cdf, norm_pdf, validate_inputs, validate_rates, OptionType, PricingError,
};

/// Full Greek set of a European option (units in the module docs).
///
/// # Examples
///
/// ```
/// use eq_options_engine::{bs_greeks, OptionType};
/// let g = bs_greeks(100.0, 100.0, 1.0, 0.05, 0.2, 0.0, OptionType::Call).unwrap();
/// assert!((g.delta - 0.6368306511756191).abs() < 1e-12);
/// assert!(g.gamma > 0.0 && g.vega > 0.0);
/// ```
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct BsGreeks {
    /// Present value, currency units.
    pub price: f64,
    /// dV/dS, dimensionless.
    pub delta: f64,
    /// d2V/dS2, per currency unit.
    pub gamma: f64,
    /// dV/dsigma, per unit of annualised vol.
    pub vega: f64,
    /// dV/dt, per year.
    pub theta: f64,
    /// dV/dr, per unit of rate.
    pub rho: f64,
    /// d2V/(dS dsigma).
    pub vanna: f64,
    /// d2V/dsigma2.
    pub volga: f64,
}

/// Analytic Black-Scholes-Merton Greeks (with continuous dividend yield).
///
/// # Errors
///
/// [`PricingError::InvalidInput`] unless `s`, `k`, `t`, `sigma` are
/// strictly positive (the Greeks are singular at the boundary).
///
/// # Examples
///
/// ```
/// use eq_options_engine::{bs_greeks, OptionType};
/// let call = bs_greeks(100.0, 100.0, 1.0, 0.05, 0.2, 0.0, OptionType::Call).unwrap();
/// let put = bs_greeks(100.0, 100.0, 1.0, 0.05, 0.2, 0.0, OptionType::Put).unwrap();
/// // Same gamma and vega for call and put:
/// assert!((call.gamma - put.gamma).abs() < 1e-15);
/// assert!((call.vega - put.vega).abs() < 1e-15);
/// ```
pub fn bs_greeks(
    s: f64,
    k: f64,
    t: f64,
    r: f64,
    sigma: f64,
    q: f64,
    option_type: OptionType,
) -> Result<BsGreeks, PricingError> {
    let (d1, d2) = d1_d2(s, k, t, r, sigma, q)?;
    let sqrt_t = t.sqrt();
    let df_q = (-q * t).exp();
    let df_r = (-r * t).exp();
    let pdf_d1 = norm_pdf(d1);

    let gamma = df_q * pdf_d1 / (s * sigma * sqrt_t);
    let vega = s * df_q * pdf_d1 * sqrt_t;
    let vanna = -df_q * pdf_d1 * d2 / sigma;
    let volga = vega * d1 * d2 / sigma;
    let common_theta = -s * df_q * pdf_d1 * sigma / (2.0 * sqrt_t);

    let (price, delta, theta, rho) = match option_type {
        OptionType::Call => {
            let nd1 = norm_cdf(d1);
            let nd2 = norm_cdf(d2);
            (
                s * df_q * nd1 - k * df_r * nd2,
                df_q * nd1,
                common_theta + q * s * df_q * nd1 - r * k * df_r * nd2,
                k * t * df_r * nd2,
            )
        }
        OptionType::Put => {
            let nmd1 = norm_cdf(-d1);
            let nmd2 = norm_cdf(-d2);
            (
                k * df_r * nmd2 - s * df_q * nmd1,
                -df_q * nmd1,
                common_theta - q * s * df_q * nmd1 + r * k * df_r * nmd2,
                -k * t * df_r * nmd2,
            )
        }
    };

    Ok(BsGreeks {
        price,
        delta,
        gamma,
        vega,
        theta,
        rho,
        vanna,
        volga,
    })
}

/// Central finite-difference Greeks for *any* pricer with the BS signature.
///
/// The pricer is a closure over `(s, k, t, r, sigma, q, option_type)`
/// returning a price (`Result`), e.g. [`crate::bs_price`] itself or a
/// closure around [`crate::crr_price`]. Central differences are used
/// everywhere; second derivatives use the three-point stencil; vanna uses
/// the four-point cross stencil. Theta is reported as `dV/dt = -dV/dT`
/// (per year).
///
/// Bump sizes: first derivatives use `rel_bump * max(|x|, 1)` with
/// `rel_bump = 1e-5` (truncation vs round-off balance for analytic
/// pricers); second derivatives use `rel_bump2 = 2e-4` (round-off scales
/// like `eps / h^2`, optimal `h ~ eps^0.25`). These match the Python
/// reference `fd_greeks` defaults.
///
/// # Errors
///
/// [`PricingError::InvalidInput`] on invalid inputs, or if `t`, `sigma`
/// or `s` is too small for the down-leg of the central bump to stay
/// inside the domain (`t - h_t <= 0`, `sigma - h_v2 <= 0`,
/// `s - h_s2 <= 0`) — this is checked up front so that a pricer closure
/// which does *not* validate its own arguments can never be evaluated at
/// a negative volatility or a negative spot. Any error raised by the
/// pricer itself is propagated unchanged.
///
/// # Examples
///
/// ```
/// use eq_options_engine::{bs_price, bs_greeks, fd_greeks, OptionType};
/// let ot = OptionType::Call;
/// let fd = fd_greeks(bs_price, 100.0, 100.0, 1.0, 0.05, 0.2, 0.0, ot).unwrap();
/// let an = bs_greeks(100.0, 100.0, 1.0, 0.05, 0.2, 0.0, ot).unwrap();
/// assert!((fd.delta - an.delta).abs() < 1e-6);
/// assert!((fd.vega - an.vega).abs() < 1e-6 * an.vega.abs().max(1.0));
/// ```
#[allow(clippy::too_many_arguments)]
pub fn fd_greeks<F>(
    pricer: F,
    s: f64,
    k: f64,
    t: f64,
    r: f64,
    sigma: f64,
    q: f64,
    option_type: OptionType,
) -> Result<BsGreeks, PricingError>
where
    F: Fn(f64, f64, f64, f64, f64, f64, OptionType) -> Result<f64, PricingError>,
{
    const REL_BUMP: f64 = 1e-5;
    const REL_BUMP2: f64 = 2e-4;

    validate_inputs(s, k, t, sigma)?;
    validate_rates(r, q)?;

    let f = |sv: f64, sig: f64, tv: f64, rv: f64| pricer(sv, k, tv, rv, sig, q, option_type);

    let h_s = REL_BUMP * s.abs().max(1.0);
    let h_v = REL_BUMP * sigma.abs().max(1.0);
    let h_t = REL_BUMP * t.abs().max(1.0);
    let h_r = REL_BUMP * r.abs().max(1.0);
    let h_s2 = REL_BUMP2 * s.abs().max(1.0);
    let h_v2 = REL_BUMP2 * sigma.abs().max(1.0);
    if t - h_t <= 0.0 {
        return Err(PricingError::InvalidInput(format!(
            "T={t} too small for a central theta bump of {h_t}"
        )));
    }
    // A central bump must stay inside the domain. Because the bumps are
    // floored at `rel_bump * 1`, a spot or vol below the bump would be
    // pushed negative by the down-leg. `bs_price` would reject that, but a
    // user-supplied pricer closure need not: without this guard the
    // returned "vega" could be a difference of prices evaluated at a
    // *negative* volatility, which is silently meaningless. Reject up
    // front instead. `h_*2 >= h_*`, so guarding the second-order bump
    // covers the first-order one too.
    if sigma - h_v2 <= 0.0 {
        return Err(PricingError::InvalidInput(format!(
            "sigma={sigma} too small for a central vega/volga bump of {h_v2}; \
             the down-leg would price at a negative volatility"
        )));
    }
    if s - h_s2 <= 0.0 {
        return Err(PricingError::InvalidInput(format!(
            "S={s} too small for a central delta/gamma bump of {h_s2}; \
             the down-leg would price at a negative spot"
        )));
    }

    let price = f(s, sigma, t, r)?;
    let delta = (f(s + h_s, sigma, t, r)? - f(s - h_s, sigma, t, r)?) / (2.0 * h_s);
    let gamma =
        (f(s + h_s2, sigma, t, r)? - 2.0 * price + f(s - h_s2, sigma, t, r)?) / (h_s2 * h_s2);
    let vega = (f(s, sigma + h_v, t, r)? - f(s, sigma - h_v, t, r)?) / (2.0 * h_v);
    let theta = -(f(s, sigma, t + h_t, r)? - f(s, sigma, t - h_t, r)?) / (2.0 * h_t);
    let rho = (f(s, sigma, t, r + h_r)? - f(s, sigma, t, r - h_r)?) / (2.0 * h_r);
    let vanna = (f(s + h_s2, sigma + h_v2, t, r)? - f(s + h_s2, sigma - h_v2, t, r)?
        - f(s - h_s2, sigma + h_v2, t, r)?
        + f(s - h_s2, sigma - h_v2, t, r)?)
        / (4.0 * h_s2 * h_v2);
    let volga =
        (f(s, sigma + h_v2, t, r)? - 2.0 * price + f(s, sigma - h_v2, t, r)?) / (h_v2 * h_v2);

    Ok(BsGreeks {
        price,
        delta,
        gamma,
        vega,
        theta,
        rho,
        vanna,
        volga,
    })
}
